### PR TITLE
[dv] Add UNR in DV methodology

### PR DESCRIPTION
**Pull request by weicaiyang**
Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

**Review by rswarbrick — CHANGES_REQUESTED** [review]
I'm not sure this is a good idea.

What happens if we waive some unreachable coverage points, then change the design to fix a bug and those coverage points become reachable? I would have thought we'd want to use UNR to do some sort of automated waiver generation, which we fed in to the eventual coverage report.

**Comment by weicaiyang** [comment]
> I'm not sure this is a good idea.
> 
> What happens if we waive some unreachable coverage points, then change the design to fix a bug and those coverage points become reachable? I would have thought we'd want to use UNR to do some sort of automated waiver generation, which we fed in to the eventual coverage report.

I think we usually review code coverage and add exclusion file when it's close to V3. Adding UNR coverage into exclusion file should be done at that period too, when design is very stable.
Designer can run UNR earlier to check the design.

My initial set-up for this flow was to automatically use this waiver in regression, but then I changed my mind due to following reasons.
1. better to have clear review process (submit to the repo) for unreachable code rather than silently exclude it
2. UNR takes 1 hours for top-level
3. require additional license
I have compared 2 options in this [slides](https://docs.google.com/presentation/d/13FwwSeBaxtHZkLMmW3TI0YGO93_BTPZZFnIpX2kQhHI/edit#slide=id.ga3bb534d6e_0_6)

**Review by rasmus-madsen — COMMENTED** [review]
I am not sure I completely follow the idea!
I get that we want to be aware of any unreachable code 

shouldn't we remove unreachable code - if not completely then with an ifdef?
Is there any point of keeping it in and excluding it?

I guess I am asking when do we want to keep unreachable code in the design?
and even if we do are we sure it is a good idea to check-in an exclusion file
if something changes and the code is no longer unreachable it will still be excluded that is not good.

**Comment by weicaiyang** [comment]
> EDIT ----
> @weicaiyang sorry my browser wasn't updated so I didn't see your response to @rswarbrick
> which I guess sort of aligns with my thoughts.
> 
> I am still wondering if there are any scenarios were we want to keep unreachable code around?

I can think of a few cases.
1. `case` statement with unreachable default branch. Here is the [example](https://docs.google.com/presentation/d/13FwwSeBaxtHZkLMmW3TI0YGO93_BTPZZFnIpX2kQhHI/edit#slide=id.ga3bb534d6e_0_42)
2. to comply with TLUL spec, we don't support a/d_param, d_sink but we still have the ports. And we only use d_opcode[0], the other 2 bits are tied to 0. [Here](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/common_cov_excl.el
) is the common UNR exclusion.
3. some primitives have more functionality than the IP needs, which could cause unreachable code

**Comment by rswarbrick** [comment]
I really don't like the idea of checking in these waivers. Apart from anything else, this is an open source project, so it probably won't quite follow the usual "get to V3; release; tape out; forget" sequence. I guess I'd be ok with this process on some release branch, but it really doesn't belong in master.

However, we can easily "do it right" by folding this in to final signoff and running the UNR flow on the thing we're actually going to tape out. Can't we just do that instead?

**Comment by rasmus-madsen** [comment]
> > EDIT ----
> > @weicaiyang sorry my browser wasn't updated so I didn't see your response to @rswarbrick
> > which I guess sort of aligns with my thoughts.
> > I am still wondering if there are any scenarios were we want to keep unreachable code around?
> 
> I can think of a few cases.
> 
> 1. `case` statement with unreachable default branch. Here is the [example](https://docs.google.com/presentation/d/13FwwSeBaxtHZkLMmW3TI0YGO93_BTPZZFnIpX2kQhHI/edit#slide=id.ga3bb534d6e_0_42)
> 2. to comply with TLUL spec, we don't support a/d_param, d_sink but we still have the ports. And we only use d_opcode[0], the other 2 bits are tied to 0. [Here](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/vcs/common_cov_excl.el) is the common UNR exclusion.
> 3. some primitives have more functionality than the IP needs, which could cause unreachable code


@weicaiyang good examples but I think I'm in line with @rswarbrick  on this.
I feel the best way to go about this is to generate exclusion only on a release branch.
I think this is anyway pretty common practice. if you check exclusions in on a dev branch - the next engineer might not be away of them and then add functionality on code that is also excluded from coverage.

Maybe we should do a release branch similar to the bronze branch - and we can do our exclusions there?

**Comment by weicaiyang** [comment]
> I really don't like the idea of checking in these waivers. Apart from anything else, this is an open source project, so it probably won't quite follow the usual "get to V3; release; tape out; forget" sequence. I guess I'd be ok with this process on some release branch, but it really doesn't belong in master.
> 
> However, we can easily "do it right" by folding this in to final signoff and running the UNR flow on the thing we're actually going to tape out. Can't we just do that instead?

Sorry, I don't quit get this "I'd be ok with this process on some release branch, but it really doesn't belong in master". If release branch is something like bronze branch, we don't need to run UNR for bronze branch as there may be some unfinished functionality and we may not even start to verify it as design sometime is not quite ready.
even we don't check in the autogen UNR exclusion file, most likely we still have to check in a coverage exclusion file to waive other types of coverage (see [Exclusion section in DV methodology](https://github.com/lowRISC/opentitan/blob/master/doc/ug/dv_methodology/index.md#exclusions)).

Is it better to have all kinds of exclusion in a file with a same review flow, rather than "some checked in, some generated on-the-fly and reviewed as sign-off item"? After we check in the exclusion and design has been updated, VCS exclusion check will fail and we have to review exclusion again, otherwise will see coverage not 100%.

Besides, it would add significant time to our nightly regression if we generate the UNR exclusion on-the-fly. It takes 1 hour for top-level to generate and it also requires additional license.

**Comment by vogelpi** [comment]
We have been discussing this in our hardware meeting and the general consensus was that while there are good reasons for doing this as outlined by @weicaiyang there are also some real dangers outlined by @rswarbrick and @rasmus-madsen we would like to avoid.

I thus suggest the following:
1. Clarify in the document that this coverage exclusion file should only be generated when going towards the corresponding development stage/milestone. Since 100% code coverage is required only at V3, it should probably happen between V2 and V3. It should be clear that it is not the right thing to add a coverage exclusion file to reach the 80% needed for V2.
2. Clarify that if any RTL changes happen to the design after the coverage exclusion file has been created, it needs to be redone and re-reviewed. This is already implicit to our process (if we say we only add this at V3) but it wouldn't hurt to re-state this in the document.
3. Clarify that all waivers have to be checked during sign-off. This not only affects coverage exclusions but also lint waivers. For lint, we have items in the design/verification checklists that explicitly require them to be checked by the technical steering committee. Something like this should also be added for the coverage exclusion file.

Some remarks to Point 2:
- I think @rasmus-madsen  meant by "I'd be ok with this process on some release branch, but it really doesn't belong in master" that we should only do this if we are very close to the release and the design is complete (no more code changes coming that could invalidate the waiver). I think this is covered by our development stages. Obviously, it does not make sense to do this on something like bronze.
- Ideally, we would generate a hash of the source code as part of the coverage exclusion file. If then at some point the source code changed (which would also put the design back into D2/V2), the coverage exclusion file would stop working. I don't know how feasible that is, but it sounds like a good thing to try.
- @weicaiyang you've stated "After we check in the exclusion and design has been updated, VCS exclusion check will fail and we have to review exclusion again, otherwise will see coverage not 100%." This sounds very much like the above point. Does this also cover the case if we change the design in such way that something that was previously covered by an exclusion file now becomes reachable? Or will we not see this because of the outdated exclusion file? I think this is the dangerous case we would like to avoid.

**Comment by rswarbrick** [comment]
I agree with what Pirmin wrote. About the master vs. release branch discussion, I was imagining a "pre-tape-out" coverage closure phase. I sort of imagined that happening on some release branch (for the tape-out), whose RTL we'd never touch again. Of course, that doesn't really work with our per-IP-block development model, where different IP blocks will get to V3 at different times, all in the same git repository.

I think I understand what you're saying about the checked-in UNR waivers. I really like the idea of including a hash of the contents of the block's `rtl` directory as a magic comment line. Then we could have some automated check that fails if the RTL is changed without regenerating (or, more likely, deleting) the UNR waivers.

**Comment by weicaiyang** [comment]
> We have been discussing this in our hardware meeting and the general consensus was that while there are good reasons for doing this as outlined by @weicaiyang there are also some real dangers outlined by @rswarbrick and @rasmus-madsen we would like to avoid.
> 
> I thus suggest the following:
> 
> 1. Clarify in the document that this coverage exclusion file should only be generated when going towards the corresponding development stage/milestone. Since 100% code coverage is required only at V3, it should probably happen between V2 and V3. It should be clear that it is not the right thing to add a coverage exclusion file to reach the 80% needed for V2.
> 2. Clarify that if any RTL changes happen to the design after the coverage exclusion file has been created, it needs to be redone and re-reviewed. This is already implicit to our process (if we say we only add this at V3) but it wouldn't hurt to re-state this in the document.
> 3. Clarify that all waivers have to be checked during sign-off. This not only affects coverage exclusions but also lint waivers. For lint, we have items in the design/verification checklists that explicitly require them to be checked by the technical steering committee. Something like this should also be added for the coverage exclusion file.

@vogelpi thanks for these valuable inputs
For 1 and 2, they make sense to me. Will add them to the document.
For 3, looks like we don't have it in the [DV sign-off checklist](https://docs.opentitan.org/doc/project/checklist/), so we just have regular PR review for coverage exclusion file. Perhaps it's good to add it in the checklist to review all EL files and cover.cfg file unless they just use common files which has been reviewed. @sriyerg WDYT?


> Some remarks to Point 2:
> 
> * I think @rasmus-madsen  meant by "I'd be ok with this process on some release branch, but it really doesn't belong in master" that we should only do this if we are very close to the release and the design is complete (no more code changes coming that could invalidate the waiver). I think this is covered by our development stages. Obviously, it does not make sense to do this on something like bronze.
> * Ideally, we would generate a hash of the source code as part of the coverage exclusion file. If then at some point the source code changed (which would also put the design back into D2/V2), the coverage exclusion file would stop working. I don't know how feasible that is, but it sounds like a good thing to try.
> * @weicaiyang you've stated "After we check in the exclusion and design has been updated, VCS exclusion check will fail and we have to review exclusion again, otherwise will see coverage not 100%." This sounds very much like the above point. Does this also cover the case if we change the design in such way that something that was previously covered by an exclusion file now becomes reachable? Or will we not see this because of the outdated exclusion file? I think this is the dangerous case we would like to avoid.

If the exclusion is in the module which has been updated, VCS checksum will fail and the exclusion will be ignored. However, if we exclude it by module and there is no update on this module, the exclusion may still be applied when some update on upper module makes it become reachable. I feel it's not supper strict and safe but we can consider it as a cross-check

**Comment by weicaiyang** [comment]
> I agree with what Pirmin wrote. About the master vs. release branch discussion, I was imagining a "pre-tape-out" coverage closure phase. I sort of imagined that happening on some release branch (for the tape-out), whose RTL we'd never touch again. Of course, that doesn't really work with our per-IP-block development model, where different IP blocks will get to V3 at different times, all in the same git repository.
> 
> I think I understand what you're saying about the checked-in UNR waivers. I really like the idea of including a hash of the contents of the block's `rtl` directory as a magic comment line. Then we could have some automated check that fails if the RTL is changed without regenerating (or, more likely, deleting) the UNR waivers.

Agree. Thanks for the clarification.

**Comment by rasmus-madsen** [comment]
thank @vogelpi for elaborating on my comments. that was indeed what I was suggesting.
I see what @rswarbrick is saying which is also true a release branch doesn't fit well with our current methodology.
I also like the idea of a generated hash that we can check.

> If the exclusion is in the module which has been updated, VCS checksum will fail and the exclusion will be ignored. However, if we exclude it by module and there is no update on this module, the exclusion may still be applied when some update on upper module makes it become reachable. I feel it's not super strict and safe but we can consider it as a cross-check

I don't like the idea of generating module based UNR from a top-level perspective. unless you generate these if something in the design changes.
if we want to add a hash on a module level it should be from a perspective of that module being a top module.
I.e no matter what stimulus is added the code is unreachable.

if the code is only unreachable due to the top. that should somehow be included in the hash - if not we are back to the risk that @rswarbrick mentioned earlier

**Comment by rswarbrick** [comment]
Hi Weicai! You say this above:

> If the exclusion is in the module which has been updated, VCS checksum will fail and the exclusion will be ignored.

That sounds an awful lot like the hash I was talking about: maybe the tooling basically does what we need already? (In which case, great!)

**Comment by weicaiyang** [comment]
> thank @vogelpi for elaborating on my comments. that was indeed what I was suggesting.
> I see what @rswarbrick is saying which is also true a release branch doesn't fit well with our current methodology.
> I also like the idea of a generated hash that we can check.
> 
> > If the exclusion is in the module which has been updated, VCS checksum will fail and the exclusion will be ignored. However, if we exclude it by module and there is no update on this module, the exclusion may still be applied when some update on upper module makes it become reachable. I feel it's not super strict and safe but we can consider it as a cross-check
> 
> I don't like the idea of generating module based UNR from a top-level perspective. unless you generate these if something in the design changes.
> if we want to add a hash on a module level it should be from a perspective of that module being a top module.
> I.e no matter what stimulus is added the code is unreachable.
> 
> if the code is only unreachable due to the top. that should somehow be included in the hash - if not we are back to the risk that @rswarbrick mentioned earlier

The hash check is enabled by default
I guess I didn't explain the "exclude by module" clearly. IP may contain some submodules as follows. When we add some exclusion for the submodule `fifo`, then update `ip_top` later to make `fifo` all reachable, the exclusion for `fifo` can pass the hash check and is still applied to coverage unless we manually remove it. It also has the same problem if excluding by instance one by one.

```
module ip_top;
  
  fifo f1;
  ...
  fifo fn;
endmodule
```

For most of IPs, we collect coverage and run UNR in IP-level. But for some IPs (like clkmgr) which heavily interacts with other modules, we will do coverage closure for them in top-level. It's also good to run UNR in top-level for them.

**Comment by weicaiyang** [comment]
> Hi Weicai! You say this above:
> 
> > If the exclusion is in the module which has been updated, VCS checksum will fail and the exclusion will be ignored.
> 
> That sounds an awful lot like the hash I was talking about: maybe the tooling basically does what we need already? (In which case, great!)

Yes, but it's not super strict. can consider it as a cross-check

**Comment by rswarbrick** [comment]
> For most of IPs, we collect coverage and run UNR in IP-level. But for some IPs (like clkmgr) which heavily interacts with other modules, we will do coverage closure for them in top-level. It's also good to run UNR in top-level for them.

Ah, I see. Well I guess the only way to do that right is to embed a hash for all the code that was included. I was thinking this would be really hard but, thinking about it, fusesoc already collects up all the cores for us, so we could probably bodge something together even if the tool doesn't support it.

**Comment by vogelpi** [comment]
Thanks @weicaiyang for the clarifications. It looks like the existing hash check is very close to what we need.

The problem you are describing with submodules is something we also haven't solved yet in the development process. For example, we don't say if IP X in L3 stage, code changes to the used primitives are impossible. Or the other way round, if a primitive changes, we currently don't require modules using it have to go back to L2/L1. This clearly has to be solved at some point as we march towards tapeout, but we shouldn't try to solve this as part of this discussion. Its worth to check if there is already an issue for this. If not, we should open one and also mention the issue around coverage waivers (also link this PR).

The point @rasmus-madsen is making is a good one, too. I might not have the right background to judge this but my understanding is that IPs verified at the top-level are an execption. We only do this for IPs that are tightly coupled to the top level and those IPs will only go into the higher development stages together with the top.

My view is that these points shouldn't block this PR, but we must document these in the doc to provide clear guidance and create an issue to keep track of it.

**Comment by rswarbrick** [comment]
Hi Weicai! I've just had a chat about this with Rasmus and we've been thinking about how this "should" work. Here is what I understand about the current situation:

- We have some blocks that are verified standalone. We want to be able to sign off verification for these blocks without freezing the rest of the design.
- We have some blocks that are verified at a chip level (to avoid having to formalise their interfaces). We will only ever sign these off once the design is frozen because it doesn't really make sense to talk about verifying them in isolation.

Given that, I suggest the following:

> **For blocks that are verified standalone:**
>
> - We run the UNR tool and generate signoffs when the block is otherwise ready to move to V3. We should run this tool on the IP block (rather than on a top-level containing the block).
> - Assuming that we embed (or the UNR tool embeds) a hash of the block's code as part of the waivers, we won't need to worry things going stale.
> - Once we've done that, we should hopefully be able to get to 100% code/FSM coverage for such a block. If not, we need to improve its testsuite.
>
> **For blocks that are verified as part of chip DV:**
>
> - It doesn't make sense to talk about closing verification of the block until the whole chip is frozen (this is the downside of "verifying everything at once")
> - So we wait until we think we are done, then make a release branch for the chip. We will never change the RTL on this branch (except for bugfixes/ECOs etc.)
> - On that branch, we run the UNR code with the top-level to generate signoffs for these blocks. We either embed a hash of the entire design, or just rely on our release process to ensure we don't change any code after this happens.

Does this sound sensible to you? I don't think that it will disallow anything that we currently think we can do.

**Comment by weicaiyang** [comment]
I have updated the PR. PTAL.

@vogelpi I also created #4183 to further discuss how primitive change would affect the life stage of the design

@rswarbrick you comment makes sense to me. For embed hash check, I think we can just use the one VCS provides, so that we don't need to invent a new one.

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [inline doc/ug/dv_methodology/index.md:488]: This means we'd have to forego the `--exclude-bypass-checks` switch.

**Comment by sriyerg** [comment]
Sorry, late to the party. 

Here are my thoughts (which I suppose is already in line with the comments already made):

- Running UNR early in the design (not checking in the waiver) is a good practice to cleanup possible dead code / make optimizations.  
- UNR is ideally after DV has reached V2 (90% coverage across all fronts already achieved through functional testing). In post-V2 (working towards V3), V3, getting the coverage up from 90% -> 100% is the hardest to achieve. It's where the law of diminishing returns kicks in. Typically, a lot of it is design exclusions (ports tied off in certain ways causing portions of the design to not be coverable), which are painstakingly hard to go through in the coverage tool. Running UNR at this stage helps reduce DV engineer's time by helping weed those out.
- The above point applies to the IPs verified at the top level as well (ones that do not have their own testbench). These IPs are reasonably simple (which is why we made the decision to not create a dedicated testbench in the first place) - hence the coverage closure is assumed to be easy enough to not require running the UNR flow. If we do need to run the UNR flow for those IPs, We dont have to wait until the full chip is frozen - we only need to wait until those IPs are D2 (coverage collection will be limited to that specific block at compile-time (planned, yet to be added), so it should be immune to the rest of the design). 
- Keeping these auto-generated exclusion files in the mainline branch is ok because if the design in the mainline branch differs from a "release branch" (if it even exists), then that would be due to minor performance / feature enhancements which need to be re-certified (see next bullet). This is the typical A0 / B0 silicon rev approach (ECO bug fixes are applied to both branches, so the designs do not differ on that aspect). 
- The design at post-V2 stage is either at D2 or D3. In either case, the design is considered stable. It is highly likely that all change to the design beyond this point are only bug-fixes. If there are feature enhancements (which requires RFC / steering committee approval), then the design version has to be bumped and the L stage (and D / V stages) have to reset for the new version. This is why 'versioning' was added to the stages tracking. Hopefully, the milestone transitions should go quicker for the new version since only those deltas need to be verified. Whatever be the case, the point is - the existing UNR generated waivers needs to be thrown away and replaced with a new one.
- Coverage exclusion waivers are super sensitive to changes in the design. If the design changes, ALL waivers have to be re-visited (the tool mandates it). The tool computes the checksum for each and every exclusion separately based on the contents on that specific line (if the line is updated, file is renamed, the code moves up or down, etc. the checksum fails and that exclusion is not applied). It really does not matter whether an exclusion is hand written or UNR-generated. This manifests as a drop in the coverage report, which IMO is sufficient enough to raise an alarm on the block's claimed quality. (tool will refuse to apply those waivers with stale checksums). The user has to re-qualify all exclusions in the tool (re-compute the checksums) and check the file in again. If the exclusion can no longer be applied, it gets dropped as a part of this process. A new PR gets made with the updated exclusions. I imagine Xcelium works the same way, I don't really know though. 
- If the design has sub-blocks that are verified and tracked with their own testbenches (example: top_earlgrey with all the IPs) the design cannot claim to be at D2 if ALL of those sub-blocks are not already at D2 (likewise for D3). So it is also unlikely that the sub-blocks change too much. 
- Prims are certainly an exception to this, since they are verified in-line as a part of the whole DUT. Prims can change either due to feature enhancements (this will require the design version to be bumped as with my previous bullet), or due to bugs. If a bug-fix affects some other block 'X' containing the same prim that was already signed off to L2.D3.V3, then we grossly overlooked something when verifying 'X' - we'd have to post-mortem it and / or revisit our entire verification methodology to see what we missed / how we missed and hopefully plug that hole. It may be possible that the fixes made to the prim is possibly not affecting X at all (maybe it is using a different parameter set). In that case, no harm done - just rerun regressions for X and leave it at L2.D3.V3. 
- UNR is run separately for the IPs and the top level. There is no reuse of exclusions from IP to top, since they are independently closed for coverage. 
- In order to fully sign-off V3, coverage has to be 100%, which means, ALL exclusions have to be **checked-in** - otherwise our regression results will not show 100% coverage.

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:488]: We don't use this switch except xbar. Xbar has a very few auto-generated exclusions, which we can review it during signoff.

**Comment by weicaiyang** [comment]
@rswarbrick @rasmus-madsen @vogelpi @sriyerg 
I have updated the PR last week. Please take a look and LMK if you have any other comment.

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [inline doc/ug/dv_methodology/index.md:472]: We already talk about 'coverage hierarchy file' in the previous section. You can just say `The same coverage hierarchy file and the exclusion files used for the simulation (which limits the coverage collection to exactly what is needed at compile-time) can be supplied to VCS UNR.'

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [inline doc/ug/dv_methodology/index.md:474]:
```suggestion
1. Generate the VCS coverage database for the block by running full regression with `--cov` switch 
```

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [inline doc/ug/dv_methodology/index.md:475]:
```suggestion
2. Launch the VCS UNR flow:
```

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [reply doc/ug/dv_methodology/index.md:488]: True, but worth mentioning this.

File: doc/ug/dv_methodology/index.md
Comment (sriyerg) [inline doc/ug/dv_methodology/index.md:489]: One sentence per line.

**Comment by weicaiyang** [comment]
@sriyerg I have updated the PR. PTAL

**Review by rswarbrick — APPROVED** [review]
This looks good to me, thanks.

File: doc/ug/dv_methodology/index.md
Comment (vogelpi) [inline doc/ug/dv_methodology/index.md:488]: Nit: missing full stop.

**Review by vogelpi — APPROVED** [review]
LGTM, thanks everybody!

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md]: Fixed. Thanks

File: doc/ug/dv_methodology/index.md
Comment (rasmus-madsen) [inline doc/ug/dv_methodology/index.md:488]:
```suggestion
It is cumbersome to go through a coverage report to manually add exclusions, but the VCS UNR flow provides a path to weed out all of the unreachables ones.
```

File: doc/ug/dv_methodology/index.md
Comment (rasmus-madsen) [inline doc/ug/dv_methodology/index.md:485]:
```suggestion
1. It's encouraged that designers run UNR to check the design in the early design stage (D1/D2), but adding an exclusion for unreachable coverage should be done between the D2 and V3 stage when the design is frozen (no feature update is allowed, except bug fix).
```

File: doc/ug/dv_methodology/index.md
Comment (rasmus-madsen) [inline doc/ug/dv_methodology/index.md:498]: @weicaiyang I think this needs rewording.
it is not so much a lack of support. it simply does not make sense to talk about unreachable code when talking functional coverage.

```suggestion
Note: VCS UNR is only relevant when closing code coverage.
```

**Review by rasmus-madsen — CHANGES_REQUESTED** [review]
@weicaiyang sorry for the delayed review.

I have a few suggestions and other than that it looks good.

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:488]: Updated. Thanks

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:485]: Done

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:498]: I feel unreachable analysis is doable for assertion. FPV tool already can support it, but it's supported in VCS.
UNR for covergroup may be possible too in the future. I kind of prefer to explicitly mention what UNR can't support. WDYT?

**Review by rasmus-madsen — APPROVED** [review]

File: doc/ug/dv_methodology/index.md
Comment (rasmus-madsen) [reply doc/ug/dv_methodology/index.md:498]: I think we are getting into semantics here.
unreachable functional coverage is to me a coverage hole

Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:498]: Agree. usually it's just due to coding issue in TB, like forget to call the sample or assign to a wrong variable. I feel tool could help to sort that out.

File: doc/ug/dv_methodology/index.md
Comment (cindychip) [inline doc/ug/dv_methodology/index.md:471]: Is it necessary to include a link to VCS UNR page? (if they have one)

**Review by cindychip — APPROVED** [review]

**Review by sriyerg — APPROVED** [review]

File: doc/ug/dv_methodology/index.md
Comment (weicaiyang) [reply doc/ug/dv_methodology/index.md:471]: I google "VCS UNR" but didn't find any official related page. I think it's only available in VCS user guide which isn't released to public.

**Pull request merged**
